File: js/src/compat/openai.ts
/** Drop-in OpenAI-style shim backed by FreeLLM.
 *
 *   // import OpenAI from "openai";
 *   import { OpenAI } from "freelm/compat";
 *   const client = new OpenAI();            // FreeLLM.fromEnv()
 *   const r = await client.chat.completions.create({
 *     model: "auto",
 *     messages: [{ role: "user", content: "hi" }],
 *   });
 *   console.log(r.choices[0].message.content);
 *
 * OpenAI-SDK constructor options ({ apiKey, baseURL, ... }) are accepted and
 * ignored — keys come from the environment / providers. `stream: true` returns
 * an async iterable of `chat.completion.chunk`-shaped objects.
 */
import { FreeLLM, FreeLLMOptions } from "../client.js";
import { ChatResponse } from "../types.js";

export interface CompatCompletion {
  id: string | null;
  object: "chat.completion";
  model: string | null;
  provider: string | null;
  choices: Array<{ index: number; message: { role: string; content: string | null; tool_calls?: any[] | null }; finish_reason: string | null }>;
  usage: { prompt_tokens: number; completion_tokens: number; total_tokens: number };
}

export interface CompatChunk {
  object: "chat.completion.chunk";
  choices: Array<{ index: number; delta: { content?: string }; finish_reason: string | null }>;
}

/** OpenAI-SDK client options, accepted for drop-in compatibility (unused),
 * plus the FreeLLM options that actually take effect. */
export interface OpenAIOptions extends FreeLLMOptions {
  apiKey?: string;
  baseURL?: string;
  organization?: string;
  project?: string;
  maxRetries?: number;
  defaultHeaders?: Record<string, string>;
  defaultQuery?: Record<string, string>;
  fetch?: unknown;
  [key: string]: any;
}

function toFreeLLMOptions(opts: OpenAIOptions): FreeLLMOptions {
  const { strategy, maxAttempts, timeout, wait, maxWait } = opts;
  const out: FreeLLMOptions = {};
  if (strategy !== undefined) out.strategy = strategy;
  if (maxAttempts !== undefined) out.maxAttempts = maxAttempts;
  if (typeof timeout === "number") out.timeout = timeout;
  if (wait !== undefined) out.wait = wait;
  if (maxWait !== undefined) out.maxWait = maxWait;
  return out;
}

function wrap(resp: ChatResponse): CompatCompletion {
  return {
    id: resp.id,
    object: "chat.completion",
    model: resp.model,
    provider: resp.provider,
    choices: resp.choices.map((c) => ({
      index: c.index,
      message: { role: c.message.role, content: c.message.content, tool_calls: c.message.tool_calls },
      finish_reason: c.finish_reason,
    })),
    usage: resp.usage,
  };
}

async function* wrapStream(deltas: AsyncGenerator<string>): AsyncGenerator<CompatChunk> {
  for await (const content of deltas) {
    yield { object: "chat.completion.chunk", choices: [{ index: 0, delta: { content }, finish_reason: null }] };
  }
}

type CreateArgs = { model?: string; messages?: any[]; stream?: boolean; stream_options?: any; [k: string]: any };

class Completions {
  constructor(private client: FreeLLM) {}

  create(args: CreateArgs & { stream: true }): Promise<AsyncGenerator<CompatChunk>>;
  create(args?: CreateArgs & { stream?: false }): Promise<CompatCompletion>;
  async create(args: CreateArgs = {}): Promise<CompatCompletion | AsyncGenerator<CompatChunk>> {
    const { model = "auto", messages = [], stream, stream_options: _so, ...rest } = args;
    if (stream) return wrapStream(this.client.stream(messages, { model, ...rest }));
    const resp = await this.client.chat(messages, { model, ...rest });
    return wrap(resp);
  }
}

class Chat {
  completions: Completions;
  constructor(client: FreeLLM) {
    this.completions = new Completions(client);
  }
}

export class OpenAI {
  chat: Chat;
  private client: FreeLLM;
  /** Accepts a FreeLLM instance, OpenAI-SDK-style options, or nothing. */
  constructor(clientOrOpts?: FreeLLM | OpenAIOptions) {
    this.client = clientOrOpts instanceof FreeLLM ? clientOrOpts : FreeLLM.fromEnv(toFreeLLMOptions(clientOrOpts ?? {}));
    this.chat = new Chat(this.client);
  }
}

File: src/freelm/compat/openai.py
"""Drop-in OpenAI-style shim.

Swap an existing OpenAI client with almost no code change::

    # from openai import OpenAI
    from freelm.compat import OpenAI

    client = OpenAI()                       # uses FreeLLM.from_env()
    r = client.chat.completions.create(
        model="auto",
        messages=[{"role": "user", "content": "hi"}],
    )
    print(r.choices[0].message.content)

OpenAI-SDK constructor arguments (``api_key``, ``base_url``, ``organization``,
...) are accepted and ignored — keys come from the environment / providers.
``stream=True`` returns an iterator of ``chat.completion.chunk``-shaped objects.
"""
from __future__ import annotations

from typing import Any, AsyncIterator, Iterator, List, Optional

from ..client import AsyncFreeLLM, FreeLLM
from ..types_compat import CompatChunk, wrap_chunk, wrap_completion  # noqa: F401  (re-export symmetry)

# OpenAI-SDK constructor kwargs we accept for drop-in compatibility but do not
# use (freelm reads keys/endpoints from its providers). ``timeout`` is forwarded
# when it's a plain number, since the semantic matches FreeLLM's.
_OPENAI_CTOR_KWARGS = frozenset(
    {
        "api_key",
        "organization",
        "project",
        "base_url",
        "websocket_base_url",
        "max_retries",
        "default_headers",
        "default_query",
        "http_client",
        "_strict_response_validation",
    }
)


def _client_kwargs(kw: dict) -> dict:
    for k in _OPENAI_CTOR_KWARGS:
        kw.pop(k, None)
    t = kw.get("timeout")
    if t is not None and not isinstance(t, (int, float)):
        kw.pop("timeout")  # httpx.Timeout objects etc. — not translatable
    return kw


def _create_kwargs(kw: dict) -> dict:
    # OpenAI passes some args freelm handles separately or ignores.
    kw.pop("stream", None)
    kw.pop("stream_options", None)
    return kw


class _Completions:
    def __init__(self, client: FreeLLM) -> None:
        self._client = client

    def create(self, *, model: str = "auto", messages: Optional[List[Any]] = None, stream: bool = False, **kw: Any):
        if stream:
            return self._create_stream(model, messages or [], _create_kwargs(kw))
        resp = self._client.chat(messages or [], model=model, **_create_kwargs(kw))
        return wrap_completion(resp)

    def _create_stream(self, model: str, messages: List[Any], kw: dict) -> Iterator[CompatChunk]:
        for delta in self._client.stream(messages, model=model, **kw):
            yield wrap_chunk(delta)


class _AsyncCompletions:
    def __init__(self, client: AsyncFreeLLM) -> None:
        self._client = client

    async def create(self, *, model: str = "auto", messages: Optional[List[Any]] = None, stream: bool = False, **kw: Any):
        if stream:
            return self._create_stream(model, messages or [], _create_kwargs(kw))
        resp = await self._client.chat(messages or [], model=model, **_create_kwargs(kw))
        return wrap_completion(resp)

    async def _create_stream(self, model: str, messages: List[Any], kw: dict) -> AsyncIterator[CompatChunk]:
        async for delta in self._client.astream(messages, model=model, **kw):
            yield wrap_chunk(delta)


class _Chat:
    def __init__(self, completions: Any) -> None:
        self.completions = completions


class OpenAI:
    """Synchronous OpenAI-compatible facade backed by FreeLLM."""

    def __init__(self, freelm: Optional[FreeLLM] = None, **kw: Any) -> None:
        self._client = freelm or FreeLLM.from_env(**_client_kwargs(kw))
        self.chat = _Chat(_Completions(self._client))

    def close(self) -> None:
        self._client.close()


class AsyncOpenAI:
    """Asynchronous OpenAI-compatible facade backed by AsyncFreeLLM."""

    def __init__(self, freelm: Optional[AsyncFreeLLM] = None, **kw: Any) -> None:
        self._client = freelm or AsyncFreeLLM.from_env(**_client_kwargs(kw))
        self.chat = _Chat(_AsyncCompletions(self._client))

    async def aclose(self) -> None:
        await self._client.aclose()
